packages/plugin-ext/src/hosted/node/plugin-ext-hosted-backend-module.ts

```diff
@@ -32,44 +32,48 @@ import { GrammarsReader } from './scanners/grammars-reader';
 
-export function bindCommonHostedBackend(bind: interfaces.Bind): void {
-    bind(HostedPluginCliContribution).toSelf().inSingletonScope();
-    bind(CliContribution).toService(HostedPluginCliContribution);
+const commonHostedConnectionModule = ConnectionContainerModule.create(({ bind, bindBackendService }) => {
+    bind(HostedPluginProcess).toSelf().inSingletonScope();
+    bind(HostedPluginSupport).toSelf().inSingletonScope();
 
-    bind(HostedPluginReader).toSelf().inSingletonScope();
+    bind(HostedPluginsManagerImpl).toSelf().inSingletonScope();
+    bind(HostedPluginsManager).toService(HostedPluginsManagerImpl);
 
+    bindContributionProvider(bind, Symbol.for(ExtPluginApiProvider));
     bind(HostedPluginServerImpl).toSelf().inSingletonScope();
     bind(HostedPluginServer).toService(HostedPluginServerImpl);
-    bind(PluginDeployerHandler).toService(HostedPluginServerImpl);
-
-    bind(HostedPluginSupport).toSelf().inSingletonScope();
-    bind(MetadataScanner).toSelf().inSingletonScope();
-    bind(HostedPluginsManager).to(HostedPluginsManagerImpl).inSingletonScope();
+    bindBackendService<HostedPluginServer, HostedPluginClient>(hostedServicePath, HostedPluginServer, (server, client) => {
+        server.setClient(client);
+        return server;
+    });
+});
 
-    bind(HostedPluginProcess).toSelf().inSingletonScope();
+export function bindCommonHostedBackend(bind: interfaces.Bind): void {
+    bind(HostedPluginCliContribution).toSelf().inSingletonScope();
+    bind(CliContribution).toService(HostedPluginCliContribution);
 
+    bind(MetadataScanner).toSelf().inSingletonScope();
+    bind(HostedPluginReader).toSelf().inSingletonScope();
     bind(BackendApplicationContribution).toService(HostedPluginReader);
 
-    bind(ConnectionHandler).toDynamicValue(ctx =>
-        new JsonRpcConnectionHandler<HostedPluginClient>(hostedServicePath, client => {
-            const server = ctx.container.get<HostedPluginServer>(HostedPluginServer);
-            server.setClient(client);
-            // FIXME: handle multiple remote connections
-            /*
-            client.onDidCloseConnection(() => server.dispose());*/
-            return server;
-        })
-    ).inSingletonScope();
+    bind(HostedPluginDeployerHandler).toSelf().inSingletonScope();
+    bind(PluginDeployerHandler).toService(HostedPluginDeployerHandler);
 
     bind(GrammarsReader).toSelf().inSingletonScope();
+
+    bind(ConnectionContainerModule).toConstantValue(commonHostedConnectionModule);
 }
 
+const hostedBackendConnectionModule = ConnectionContainerModule.create(({ bind }) => {
+    bindContributionProvider(bind, Symbol.for(HostedPluginUriPostProcessorSymbolName));
+    bind(HostedInstanceManager).to(NodeHostedPluginRunner).inSingletonScope();
+});
+
 export function bindHostedBackend(bind: interfaces.Bind): void {
     bindCommonHostedBackend(bind);
+    bind(ConnectionContainerModule).toConstantValue(hostedBackendConnectionModule);
```

Comment:
hello, simple question, is that ConnectionContainerModule is handling the destroy of the instances associated to the scope ? (closing the browser tab is destroying the instances of this scope ) ?

Reply:
yes, it should, instances only exists within connection and can access each other (be injected)

Reply:
> closing the browser tab is destroying the instances of this scope

@svenefftinge it was my assumption. It could be good to implement, otherwise it could be hard to clean up without explicitly binding backend service.

Reply:
yes I assume we may have leaks on many items if instances are not destroyed/not notified ?